### PR TITLE
Add Iterator.Count

### PR DESCRIPTION
**Pull request by shargon**

### CODE REVIEW

File: src/neo/SmartContract/ApplicationEngine.Iterator.cs [hunk 1]
```diff
@@ -38,6 +44,23 @@ internal protected static bool IteratorNext(IIterator iterator)
...
+        /// <summary>
+        /// The implementation of System.Iterator.Count.
+        /// Returns the number of entries in the collection.
+        /// </summary>
+        /// <param name="iterator">The iterator to be counted.</param>
+        /// <returns>Seek the iterator to the last entry and return the number of entries.</returns>
+        internal protected int IteratorCount(IIterator iterator)
+        {
+            int count = 0;
+            while (iterator.Next())
```
Comment (Liaojinghui) [inline src/neo/SmartContract/ApplicationEngine.Iterator.cs:56]: Won't this change the iterator position?

**Comment by roman-khimov** [comment]
What's the benefit compared to contract making a series of `iterator.Next()` calls by itself?

File: src/neo/SmartContract/ApplicationEngine.Iterator.cs (hunk 1 above)
Comment (shargon) [reply src/neo/SmartContract/ApplicationEngine.Iterator.cs:56]: Yes

**Comment by shargon** [comment]
> What's the benefit compared to contract making a series of iterator.Next() calls by itself?

Less expensive

**Comment by roman-khimov** [comment]
Each `iterator.Next()` costs 32768, JMPIF/INC/PUSH/DUP are 2/4/1/2 (STLOC/LDLOC are 2/2 if needed), so additional code to iterate and count maybe costs around 32 per iteration which is less than 0.1% overhead.

File: src/neo/SmartContract/ApplicationEngine.Iterator.cs (hunk 1 above)
Comment (Qiao-Jin) [reply src/neo/SmartContract/ApplicationEngine.Iterator.cs:56]: Is this by design, to change iterator position?

**Comment by erikzhang** [comment]
After calling `Count()`, the position of the `Iterator` is not restored?

**Comment by shargon** [comment]
If we want to restore the position, we need to add the method in the interface

**Comment by Qiao-Jin** [comment]
We can restore the position by searching & comparing keys with the initial key value
But honestly I suggest changing IEnumerator property to some other data strctures i.e. List

**Pull request closed without merging**